enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,317 @@
+
+### Non-Goals
+
+This enhancement applies only to the CoreDNS Operator shipped as part of OpenShift. Anything outside that scope would be not applicable to this enhancement including the CoreDNS static pods managed by the `machine-config-operator`.
+
+## Proposal
+
+This proposal covers changes to two APIs. The first is a change to 
+`UpstreamResolvers`.  To enable a cluster administrator to configure 
+DNS-over-TLS, the `UpstreamResolvers` type will get two new optional members:
+
+`CABundle` references a ConfigMap that contains a single CA certificate 
+or a CA bundle.
+
+`ServerName` contains an FQDN to validate when establishing the connection. 
+This is relevant in particular with cases where a customer has multiple 
+upstreams behind one serving certificate (e.g. `1.1.1.1` and `2.2.2.2` 
+behind `dns.myorg.com`) as well as indicating what server we want to connect to.
+
+Since `UpstreamResolvers` doesn't allow for configuration of a custom zone, 
+it's assumed that the same TLS configuration will apply to _n_ `Upstream` 
+objects. 
+
+```go
+type UpstreamResolvers struct {
+  // ...<snip>
+
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). Populating a configmap name in this field
+  // effectively enables DNS-over-TLS. If this is empty or the configmap
+  // referenced does not exist, DNS-over-TLS will not be enabled.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key
+  // 2. The value must be a PEM encoded certificate or certificate bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  // 4. The ServerName value must be supplied when CABundle is configured.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
+
+  // serverName is the upstream server to connect to for DNS resolution.
+  // This is required when a CABundle is supplied. Populating the
+  // ServerName without a CABundle configured will have no effect.
+  //
+  // +optional
+  // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}([a-zA-Z0-9-\p{L}]){2,63})$`
+  ServerName string `json:"serverName,omitempty"`
+}
+```
+
+`Upstream` will get one new optional member:
+
+`EnableTLS` which allows an individual upstream to be opted into using
+DNS-over-TLS. This will be useful in cases where not all upstreams support
+DNS-over-TLS.
+
+```go
+type Upstream struct {
+  // ...snip
+
+  // enableTls allows cluster administrators to opt-in to using the TLS config in
+  // spec.upstreamResolvers for upstreams that support DNS-over-TLS.
+  //
+  // +optional
+  // +kubebuilder:validation:Enum=true;false
+  // +kubebuiler:default=false
+  EnableTLS string `json:"enableTls,omitempty"`
```

Comment:
nothing is a bool, how about a transport enum with TLS and ClearText

Reply:
I like that idea! Implemented locally.